patches/api/0453-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,1090 @@
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/BasicCommand.java b/src/main/java/io/papermc/paper/command/brigadier/BasicCommand.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..88ed7ee00b6cf9d9aa8f1e6b1406f41f190563d0
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/BasicCommand.java
+@@ -0,0 +1,37 @@
++package io.papermc.paper.command.brigadier;
++
++import java.util.Collection;
++import java.util.Collections;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * This represents a simple command implementation that is wrapped
++ * around brigadier.
++ */
++@ApiStatus.Experimental
++@FunctionalInterface
++public interface BasicCommand {
++
++    /**
++     * Executes the command with the given {@link CommandSourceStack} and arguments.
++     *
++     * @param commandSourceStack The commandSourceStack of the command.
++     * @param args The arguments of the command.
++     * @return The result of the command execution.
++     */
++    @ApiStatus.OverrideOnly
++    int execute(@NotNull CommandSourceStack commandSourceStack, @NotNull String[] args);
++
++    /**
++     * Suggests possible completions for the given command {@link CommandSourceStack} and arguments.
++     *
++     * @param commandSourceStack The commandSourceStack of the command.
++     * @param args The arguments of the command.
++     * @return A collection of suggestions
++     */
++    @ApiStatus.OverrideOnly
++    default @NotNull Collection<String> suggest(final @NotNull CommandSourceStack commandSourceStack, final @NotNull String[] args) {
++        return Collections.emptyList();
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..fee4a3a1ab20aa611b5ad13bc30127833a6a0ed9
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/command/brigadier/CommandSourceStack.java
+@@ -0,0 +1,54 @@
++package io.papermc.paper.command.brigadier;
++
++import com.mojang.brigadier.context.CommandContext;
++import com.mojang.brigadier.exceptions.CommandSyntaxException;
++import io.papermc.paper.command.brigadier.argument.ArgumentResolver;
++import org.bukkit.Location;
++import org.bukkit.World;
++import org.bukkit.command.CommandSender;
++import org.bukkit.entity.Entity;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++/**
++ * Represents a vanilla command source stack which is used
++ * for command execution.
++ */
++@ApiStatus.NonExtendable
++@ApiStatus.Experimental
++public interface CommandSourceStack {
++
++    /**
++     * Gets the location that this command is being executed at.
++     * @return location
++     */
++    @NotNull Location getLocation();
```

Comment:
I dont think this should be nonnull, even if vanilla will have a default one set

Reply:
Confused, why here? Where would we return null then?

Reply:
The console

Reply:
But any positioned-based commands will still execute at 0,0,0... right? So the location isn't null tho... plus I don't understand how we would be able to distinguish between literal 0,0,0 or default 0,0,0

Reply:
It just doesnt make sense for the console to have a location (unless specified), it is not bound to any world or location. Just because vanilla has or needs one doesn't mean we do in our api. Distuingishing it is easy by just going to where it is initially created for the console with the default value

Reply:
Resolved in call, decided to go with an isConsole() check on sender instead.